### PR TITLE
Allow tiberius to be run on multiple runtimes

### PR DESCRIPTION
**Pull request by esheppa**
Main changes
* Refactored to a core crate plus a crate for each runtime
* Added a macro to tests that can test on all runtimes

I attempted to make the use of the runtime based crates less confusing by having both a local `Client` and a local `ClientBuilder`, such that `tiberius_<rt>::Client::builder()` returns a local `ClientBuilder` rather than one from the core crate. It would be quite easy to remove this if this is not desired.

### CODE REVIEW

**Comment by yoshuawuyts** [comment]
Was asked by @pimeys to chime in here, so hi!

Another direction that might be worth exploring is instead of exposing separate crates for async-std and tokio, is to move towards an io-agnostic design. Instead of creating a connection inside Tiberius, tiberius would take an `impl AsyncRead + AsyncWrite` and run queries over that. This is the design we went for with `async-h1`, and the resulting API (on master) for creating a client is now this:

```rust
use async_std::net::TcpStream;
use http_types::{Method, Request, Url};

#[async_std::main]
async fn main() -> http_types::Result<()> {
    let stream = TcpStream::connect("127.0.0.1:8080").await?;

    let peer_addr = stream.peer_addr()?;
    println!("connecting to {}", peer_addr);
    let url = Url::parse(&format!("http://{}/foo", peer_addr))?;

    let req = Request::new(Method::Get, url);
    let res = async_h1::connect(stream.clone(), req).await?;
    println!("{:?}", res);

    Ok(())
}
```

This design works with any runtime that uses the `futures::{AsyncRead,AsyncWrite}` traits. The biggest change is that connection pooling would become an external concern; but if a good library exists for that it'd be a few extra lines to setup overall. From looking at how Tiberius operates this seems like it could be a direction that could work, and would help greatly reduce the surface area of the library and amount of changes that would be needed in the future.

Hope this was a useful contribution; thanks heaps!

**Comment by esheppa** [comment]
Hi @yoshuawuyts , thanks for the review (and also your excellent blog!). My intention is to model this similarly to the approach taken in `async-h1` and also `async-native-tls`. 

The main structure being: 
* tiberius (core crate, no runtime dependencies)
* tiberius-\<runtime\> for each runtime

Most of the code in the runtime specific crates just delegates to the underlying `tiberius::Client` and exists primarily for ergonomic purposes (however it does mean there is some duplication, so potentially this should be removed if it doesn't meet a cost-benefit trade off).  Also it would be possible for a user to just use the core crate and then implement the [`connector`](https://github.com/esheppa/tiberius/blob/f7be1a3e06e8b02b814ca8a6f0f36d6aeac7850c/tiberius-asyncstd/src/lib.rs#L234) function themselves. There is also another function [`find_tcp_port`](
https://github.com/esheppa/tiberius/blob/f7be1a3e06e8b02b814ca8a6f0f36d6aeac7850c/tiberius-asyncstd/src/lib.rs#L258) which does a lookup via the SQL Browser process (windows only) to connect to a named instance, but the TDS protocol requires this part to use UDP. This could also be implemented in a project that depends only on the core crate but it is a bit more finicky.

I can think of the following options to reduce or eliminate the need for the runtime specific crates:
* `find_tcp_port` could be implemented without async, and then potentially cached somewhere, thus only risking blocking on the first connection (this could then be documented and turned on/off optionally)
* we could provide the sample implementations of `connector` and `find_tcp_port` for each runtime in the examples folder rather than publishing a crate for each runtime
* potentially the `async-std` and `smol` crates can be merged
* the runtime specific crates could be trimmed down to just `connector` and `find_tcp_port` functions and a few type aliases, and then the docs could include an example of how to use the core and runtime-specific crates together

Do you or @pimeys have any preference in terms of these options - or potentially some altogether different approaches?

File: tiberius-asyncstd/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    }
+
+    /// Sets the authentication method.
+    pub fn authentication(&mut self, auth: tiberius::AuthMethod) {
+        self.inner.authentication(auth)
+    }
+}
+
+
+fn connector(addr: String, instance_name: Option<String>) -> Pin<Box<dyn Future<Output = tiberius::Result<net::TcpStream>>>>
```
[the remark this answers is not in the document]
Comment (esheppa) [reply tiberius-asyncstd/src/lib.rs]: good call, I've pushed a commit replacing all the `Pin<Box<dyn Future>>`'s

File: tiberius-smol/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,286 @@
+//! The Tiberius Microsot SQL Server driver implemented on the smol runtime
```
[the remark this answers is not in the document]
Comment (esheppa) [reply tiberius-smol/src/lib.rs]: I've renamed the async-std crate to `tiberius-async-std`

**Comment by yoshuawuyts** [comment]
Gah, I wanted to write a long-form reply to this and consult with @pimeys first. But today was a bit hectic and didn't get around to that, but I also I didn't want this to slip until Monday so here's my rough take.

Intuitively I lean towards exploring how well an io-less model could work, and creating examples for that. That seems like the base for the other approaches as well, and would allow us to pivot to runtime-specific crates fairly easily if it doesn't work out.

For the UDP lookup: maybe we could define an API that takes an async closure in which the lookup can be done? Or perhaps marking something as blocking so it can be put in a `Task::spawn_blocking` could work as well.

Either way; thanks for caring about my opinion, and thanks heaps for this!

**Comment by esheppa** [comment]
I've had some further thoughts, incorporating the information from this thread:
* We could move the address and instance_name parameters entirely out of the connection string / `ClientBuilder` - this is a little unconventional perhaps, but then it enables the following:
* When running on Linux or Windows (without named instances) users can separately set up TCP connections and then pass them to `ClientBuilder::new` to get a `Client`.  (whereas currently `tiberius` calls the `connector` function to set up the connection)
* On Windows with named instances it is a little more complicated, however it may be the case that connecting to named instances is a lower proportion of use cases so it may be acceptable for it to be slightly less ergonomic, but we could use the async closure API as suggested, and have an example which can be copy pasted into projects
* as another alternative, `tiberius` could even export a basic synchronous version of `find_tcp_port` with the intention that crate consumers run it at startup or in a `run_blocking`. 

@pimeys & @yoshuawuyts, if the above is palatable, then potentially we don't need the runtime specific crates and can simply provide information on how to set up a connection by having an example for each runtime.

**Comment by pimeys** [comment]
> We could move the address and instance_name parameters entirely out of the connection string / `ClientBuilder` - this is a little unconventional perhaps, but then it enables the following:
> When running on Linux or Windows (without named instances) users can separately set up TCP connections and then pass them to `ClientBuilder::new` to get a `Client`.  (whereas currently `tiberius` calls the `connector` function to set up the connection)

I'd say this might be worth a try.

> On Windows with named instances it is a little more complicated, however it may be the case that connecting to named instances is a lower proportion of use cases so it may be acceptable for it to be slightly less ergonomic, but we could use the async closure API as suggested, and have an example which can be copy pasted into projects

Exactly. I'm also leaning on this, especially when I've interviewed some potential Prisma users who'd like to use SQL Server for their projects. Almost always the answer is Azure SQL, when asked how would they use SQL Server. They don't really need named instances.

> as another alternative, `tiberius` could even export a basic synchronous version of `find_tcp_port` with the intention that crate consumers run it at startup or in a `run_blocking`.

Why not directly an asynchronous version?

> @pimeys & @yoshuawuyts, if the above is palatable, then potentially we don't need the runtime specific crates and can simply provide information on how to set up a connection by having an example for each runtime.

This would be sweet. We'd need some good examples though if we take this path.

File: runtimes-macro/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+#[derive(Debug, FromMeta)]
+struct MacroArgs {
+    #[darling(default)]
+    connection_string: Option<String>,
+}
+
+#[proc_macro_attribute]
+pub fn test_on_runtimes(args: proc_macro::TokenStream, input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    // let attributes = parse_macro_input!(attr as syn::Attribute);
```
Comment (pimeys) [inline runtimes-macro/src/lib.rs:12]: You might want to remove these comments.

File: runtimes-macro/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    let args = match MacroArgs::from_list(&attr_args) {
+        Ok(v) => v,
+        Err(e) => { return proc_macro::TokenStream::from(e.write_errors()); }
+    };
+
+    let func = syn::parse_macro_input!(input as syn::ItemFn);
+
+    let conn_str_ident_str = args.connection_string.unwrap_or_else(|| "CONN_STR".into());
+
```
Comment (pimeys) [inline runtimes-macro/src/lib.rs:25]: Whitespace.

File: runtimes-macro/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    let conn_str_ident = proc_macro2::Ident::new(&conn_str_ident_str, proc_macro2::Span::call_site());
+
+    // todo: ensure func is async
```
Comment (pimeys) [inline runtimes-macro/src/lib.rs:29]: Remove.

File: src/client/builder.rs [hunk 6]
```diff
@@ -1,29 +1,37 @@
...
+    /// and `AsyncWrite` but it is up to the user to ensure that it does
+    /// this via the TCP protocol.
+    pub fn new(
+    ) -> ClientBuilder {
+        ClientBuilder {
             host: None,
             port: None,
             database: None,
-            #[cfg(windows)]
+            //#[cfg(windows)]
```
Comment (pimeys) [inline src/client/builder.rs:34]: ?

File: src/client/connection.rs [hunk 7]
```diff
@@ -44,14 +33,14 @@ use winauth::{windows::NtlmSspiBuilder, NextBytes};
 ///
 /// [`Client`]: struct.Encode.html
 /// [`Packet`]: ../protocol/codec/struct.Packet.html
-pub(crate) struct Connection {
-    transport: Framed<MaybeTlsStream, PacketCodec>,
+pub(crate) struct Connection<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> {
```
Comment (pimeys) [inline src/client/connection.rs:36]: You might want to `use futures::{AsyncRead, AsyncWrite}` to save valuable horizontal space.

File: src/client/connection.rs [hunk 8]
```diff
@@ -44,14 +33,14 @@ use winauth::{windows::NtlmSspiBuilder, NextBytes};
...
+    transport: Framed<MaybeTlsStream<S>, PacketCodec>,
     flushed: bool,
     context: Context,
     buf: BytesMut,
 }
 
-impl Debug for Connection {
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> Debug for Connection<S> {
```
Comment (pimeys) [inline src/client/connection.rs:43]: You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.

File: src/client/connection.rs [hunk 9]
```diff
@@ -62,42 +51,34 @@ impl Debug for Connection {
...
-}
-
 enum LoginResult {
     Ok,
     #[cfg(windows)]
     Windows(Box<dyn NextBytes>),
 }
 
-impl Connection {
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> Connection<S> {
```
Comment (pimeys) [inline src/client/connection.rs:60]: You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.

File: src/client/connection.rs [hunk 10]
```diff
@@ -62,42 +51,34 @@ impl Debug for Connection {
...
     where
-        T: ToSocketAddrs,
     {
-        let mut addr = tokio::net::lookup_host(addr).await?.next().ok_or_else(|| {
-            io::Error::new(io::ErrorKind::NotFound, "Could not resolve server host.")
-        })?;
-
-        if let Some(ref instance_name) = opts.instance_name {
-            addr = Self::find_tcp_port(addr, instance_name).await?;
+        // need to do a create context in hre
```
Comment (pimeys) [inline src/client/connection.rs:68]: Maybe not so useful comment.

File: src/client/connection.rs [hunk 11]
```diff
@@ -388,70 +369,10 @@ impl Connection {
...
-
-        addr.set_port(port);
-
-        Ok(addr)
-    }
 }
 
-impl Stream for Connection {
+
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> Stream for Connection<S> {
```
Comment (pimeys) [inline src/client/connection.rs:375]: You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.

File: src/client/connection.rs [hunk 12]
```diff
@@ -468,7 +389,7 @@ impl Stream for Connection {
     }
 }
 
-impl AsyncRead for Connection {
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> futures::AsyncRead for Connection<S> {
```
Comment (pimeys) [inline src/client/connection.rs:392]: You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.

File: src/client/connection.rs [hunk 13]
```diff
@@ -508,7 +429,7 @@ impl AsyncRead for Connection {
     }
 }
 
-impl SqlReadBytes for Connection {
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> SqlReadBytes for Connection<S> {
```
Comment (pimeys) [inline src/client/connection.rs:432]: You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.

File: src/client/tls.rs [hunk 14]
```diff
@@ -10,45 +10,32 @@ use futures::ready;
...
 #[cfg(feature = "tls")]
-use tokio_tls::TlsStream;
+use async_native_tls::TlsStream;
 #[cfg(feature = "tls")]
 use tracing::{event, Level};
 
 /// A wrapper to handle either TLS or bare connections.
-pub(crate) enum MaybeTlsStream {
-    Raw(TcpStream),
+pub(crate) enum MaybeTlsStream<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> {
```
Comment (pimeys) [inline src/client/tls.rs:22]: You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.

File: src/client/tls.rs [hunk 15]
```diff
@@ -10,45 +10,32 @@ use futures::ready;
...
+    Raw(S),
     #[cfg(feature = "tls")]
-    Tls(TlsStream<TlsPreloginWrapper<TcpStream>>),
+    Tls(TlsStream<TlsPreloginWrapper<S>>),
 }
 
-impl MaybeTlsStream {
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> MaybeTlsStream<S> {
```
Comment (pimeys) [inline src/client/tls.rs:28]: You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.

File: src/client/tls.rs [hunk 16]
```diff
@@ -10,45 +10,32 @@ use futures::ready;
...
-impl AsyncRead for MaybeTlsStream {
-    unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [MaybeUninit<u8>]) -> bool {
-        match self {
-            MaybeTlsStream::Raw(s) => s.prepare_uninitialized_buffer(buf),
-            #[cfg(feature = "tls")]
-            MaybeTlsStream::Tls(s) => s.prepare_uninitialized_buffer(buf),
-        }
-    }
-
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> futures::AsyncRead for MaybeTlsStream<S> {
```
Comment (pimeys) [inline src/client/tls.rs:38]: You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.

File: src/client/tls.rs [hunk 17]
```diff
@@ -62,7 +49,7 @@ impl AsyncRead for MaybeTlsStream {
     }
 }
 
-impl AsyncWrite for MaybeTlsStream {
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> futures::AsyncWrite for MaybeTlsStream<S> {
```
Comment (pimeys) [inline src/client/tls.rs:52]: You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.

File: src/client/tls.rs [hunk 18]
```diff
@@ -132,7 +119,7 @@ impl<S> TlsPreloginWrapper<S> {
 }
 
 #[cfg(feature = "tls")]
-impl<S: AsyncRead + AsyncWrite + Unpin> AsyncRead for TlsPreloginWrapper<S> {
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> futures::AsyncRead for TlsPreloginWrapper<S> {
```
Comment (pimeys) [inline src/client/tls.rs:122]: You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.

File: src/client/tls.rs [hunk 19]
```diff
@@ -195,7 +182,7 @@ impl<S: AsyncRead + AsyncWrite + Unpin> AsyncRead for TlsPreloginWrapper<S> {
 }
 
 #[cfg(feature = "tls")]
-impl<S: AsyncRead + AsyncWrite + Unpin> AsyncWrite for TlsPreloginWrapper<S> {
+impl<S: futures::AsyncRead + futures::AsyncWrite + Unpin + Send> futures::AsyncWrite for TlsPreloginWrapper<S> {
```
Comment (pimeys) [inline src/client/tls.rs:185]: You might want to use futures::{AsyncRead, AsyncWrite} to save valuable horizontal space.

File: src/tds/stream/prepared.rs [hunk 20]
```diff
@@ -0,0 +1,84 @@
+use super::ReceivedToken;
```
Comment (pimeys) [inline src/tds/stream/prepared.rs:1]: I remember removing this file recently. Maybe a merge conflict gone wrong. We don't really need it and it just causes confusion by existing.

File: src/client/builder.rs (hunk 6 above)
Comment (esheppa) [reply src/client/builder.rs:31]: awks! commented these out to verify if the named-instance stuff could compile

**Pull request merged**